File: Dynamic Programming/C#/1699 제곱수의 합.cs
using System;

class Program
{
    static void Main() {
        string s = Console.ReadLine();
        int n = int.Parse(s);
        int[] dp = new int[100001];
        
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 2;
        dp[3] = 3;
        
        for(int i = 4; i <= n ; i++) {
            
            if((int)(Math.Pow((int)(Math.Pow(i, 0.5)), 2)) == i) dp[i] = 1;
            else {
                dp[i] = dp[i-1] + 1 ;
                
                for(int j = 1; j*j <= i; j++) {
                    
                    dp[i] = Math.Min(dp[i], dp[i-j*j] + dp[j*j]);
                }
            }
            
        }
        
        Console.WriteLine(dp[n]);
    }
}

// http://boj.kr/0b0efbe256f049d09ed4dfad81557e50

File: Python/1699 제곱수의 합.py
n = int(input())
dp = [0 for i in range(100001)]

dp[1] = 1
dp[2] = 2
dp[3] = 3

for i in range(4, n+1) :
    if i == int(int(i**0.5)**2) :
        dp[i] = 1
    else :
        dp[i] = dp[i-1] + 1
        j = 1
        while i > j*j :
            dp[i] = min(dp[i], dp[i-j*j] + dp[j*j])
            j += 1
        
print(dp[n])


# http://boj.kr/34ebc7734b4544ff90601705493fb67c
